From 48bbfabdb88508783d72c08567a111ecd2991e68 Mon Sep 17 00:00:00 2001
From: FrogAi <91348155+FrogAi@users.noreply.github.com>
Date: Wed, 24 May 2023 17:26:16 -0700
Subject: [PATCH] Fast boot using prebuilt file

Added prebuilt functionality that removes itself whenever the user installs updates to prevent installation failures.
---
 common/spinner.py                            |  4 +++-
 launch_chffrplus.sh                          |  8 ++++++++
 selfdrive/thermald/thermald.py               |  5 +++++
 selfdrive/ui/qt/offroad/software_settings.cc |  6 ++++++
 selfdrive/ui/qt/widgets/offroad_alerts.cc    | 11 ++++++++++-
 5 files changed, 32 insertions(+), 2 deletions(-)

diff --git a/common/spinner.py b/common/spinner.py
index 43d4bb2cc..64c9d2db3 100644
--- a/common/spinner.py
+++ b/common/spinner.py
@@ -47,6 +47,8 @@ if __name__ == "__main__":
   import time
   with Spinner() as s:
     s.update("Spinner text")
-    time.sleep(5.0)
+    # time.sleep(5.0)
+    # PFEIFER - FB {{
+    # }} PFEIFER - FB
   print("gone")
   time.sleep(5.0)
diff --git a/launch_chffrplus.sh b/launch_chffrplus.sh
index 7578c0296..edb65768e 100755
--- a/launch_chffrplus.sh
+++ b/launch_chffrplus.sh
@@ -82,6 +82,14 @@ function launch {
   # write tmux scrollback to a file
   tmux capture-pane -pq -S-1000 > /tmp/launch_log
 
+  # PFEIFER - FB {{
+  # Fastboot Acknowledgements:
+  # Fastboot changes were (nearly) directly copied from frogpilot. Thanks FrogAi! https://github.com/frogAi/FrogPilot
+  if [ -f "$BASEDIR/prebuilt" ]; then
+    python /data/openpilot/common/spinner.py &
+  fi
+  # }} PFEIFER - FB
+
   # start manager
   cd selfdrive/manager
   ./build.py && ./manager.py
diff --git a/selfdrive/thermald/thermald.py b/selfdrive/thermald/thermald.py
index 7011ff0a9..5fc0671f4 100755
--- a/selfdrive/thermald/thermald.py
+++ b/selfdrive/thermald/thermald.py
@@ -450,6 +450,11 @@ def thermald_thread(end_event, hw_queue) -> None:
     count += 1
     should_start_prev = should_start
 
+    # PFEIFER - FB {{
+    # Create the prebuilt file if it doesn't exist
+    if not os.path.isfile('/data/openpilot/prebuilt'):
+      os.system(f"touch {'/data/openpilot/prebuilt'}")
+    # }} PFEIFER - FB
 
 def main():
   hw_queue = queue.Queue(maxsize=1)
diff --git a/selfdrive/ui/qt/offroad/software_settings.cc b/selfdrive/ui/qt/offroad/software_settings.cc
index 15c022db9..7dff70d33 100644
--- a/selfdrive/ui/qt/offroad/software_settings.cc
+++ b/selfdrive/ui/qt/offroad/software_settings.cc
@@ -6,6 +6,9 @@
 
 #include <QDebug>
 #include <QLabel>
+// PFEIFER - FB {{
+#include <QProcess>
+// }} PFEIFER - FB
 
 #include "common/params.h"
 #include "common/util.h"
@@ -44,6 +47,9 @@ SoftwarePanel::SoftwarePanel(QWidget* parent) : ListWidget(parent) {
   // install update btn
   installBtn = new ButtonControl(tr("Install Update"), tr("INSTALL"));
   connect(installBtn, &ButtonControl::clicked, [=]() {
+    // PFEIFER - FB {{
+    QProcess::execute("rm -f /data/openpilot/prebuilt"); // Remove the prebuilt file when installing updates
+    // }} PFEIFER - FB
     installBtn->setEnabled(false);
     params.putBool("DoReboot", true);
   });
diff --git a/selfdrive/ui/qt/widgets/offroad_alerts.cc b/selfdrive/ui/qt/widgets/offroad_alerts.cc
index 74ece36d1..ebd3c1b8d 100644
--- a/selfdrive/ui/qt/widgets/offroad_alerts.cc
+++ b/selfdrive/ui/qt/widgets/offroad_alerts.cc
@@ -8,6 +8,9 @@
 #include <QHBoxLayout>
 #include <QJsonDocument>
 #include <QJsonObject>
+// PFEIFER - FB {{
+#include <QProcess>
+// }} PFEIFER - FB
 
 #include "common/util.h"
 #include "system/hardware/hw.h"
@@ -46,7 +49,13 @@ AbstractAlert::AbstractAlert(bool hasRebootBtn, QWidget *parent) : QFrame(parent
     QPushButton *rebootBtn = new QPushButton(tr("Reboot and Update"));
     rebootBtn->setFixedSize(600, 125);
     footer_layout->addWidget(rebootBtn, 0, Qt::AlignBottom | Qt::AlignRight);
-    QObject::connect(rebootBtn, &QPushButton::clicked, [=]() { Hardware::reboot(); });
+    // QObject::connect(rebootBtn, &QPushButton::clicked, [=]() { Hardware::reboot(); });
+    // PFEIFER - FB {{
+    QObject::connect(rebootBtn, &QPushButton::clicked, [=]() {
+      QProcess::execute("rm -f /data/openpilot/prebuilt"); // Remove the prebuilt file when installing updates
+      Hardware::reboot();
+    });
+    // }} PFEIFER - FB
   }
 
   setStyleSheet(R"(
-- 
2.39.2

